xwiki-platform-core/xwiki-platform-image/xwiki-platform-image-processing/xwiki-platform-image-processing-plugin/src/main/java/com/xpn/xwiki/plugin/image/ImagePlugin.java

```diff
@@ -367,33 +367,85 @@ private XWikiAttachment shrinkImage(XWikiAttachment attachment, int requestedWid
-            {
-                // Ignore the requested width. Check the requested height.
-                if (requestedHeight > 0 && requestedHeight < currentHeight) {
-                    // Reduce the height, keeping aspect ratio.
-                    width = (int) (requestedHeight * aspectRatio);
-                    height = requestedHeight;
-                }
-            } else {
-                // Ignore the requested height. Reduce the width, keeping aspect ratio.
-                width = requestedWidth;
-                height = (int) (requestedWidth / aspectRatio);
+        double aspectRatio = (double) currentWidth / (double) currentHeight;
+        // Ignore the width if it is not given or too large, i.e., larger than the current width or larger than the
+        // width derived from the requested height.
+        if (requestedWidth <= 0 || requestedWidth >= currentWidth
+            || (requestedHeight > 0 && requestedWidth > (int) (requestedHeight * aspectRatio)))
+        {
+            // Ignore the requested width. Check the requested height.
+            if (requestedHeight > 0 && requestedHeight < currentHeight) {
+                // Reduce the height, keeping aspect ratio.
+                width = (int) (requestedHeight * aspectRatio);
+                height = requestedHeight;
             }
         } else {
-            // Reduce both width and height, possibly loosing aspect ratio.
-            width = Math.min(requestedWidth, currentWidth);
-            height = Math.min(requestedHeight, currentHeight);
+            // Ignore the requested height. Reduce the width, keeping aspect ratio.
+            width = requestedWidth;
+            height = (int) (requestedWidth / aspectRatio);
+        }
+
+        return new int[] { width, height };
+    }
+
+    /**
+     * Reduce the image of dimension {@code currentWidth*currentHeight} so that its fit a box of size
+     * {@code requestedWidth*requestedHeight}. If the requested box is larger than {@code currentWidth*currentHeight},
+     * it is rescaled to fit it.
+     *
+     * @param currentWidth the current width of the image
+     * @param currentHeight the current height of the image
+     * @param requestedWidth the requested width of the image
+     * @param requestedHeight the requested height of the image
+     * @return the new dimensions of the image, with a new aspect ratio
+     */
+    private int[] reduceImageDimensionsForceAspectRatio(int currentWidth, int currentHeight, int requestedWidth,
+        int requestedHeight)
+    {
+        int width = currentWidth;
+        int height = currentHeight;
+        double requestedVerticalRatio = (double) requestedWidth / (double) requestedHeight;
+        double requestedHorizontalRatio = (double) requestedHeight / (double) requestedWidth;
+
+        double originalVerticalRatio = (double) currentWidth / (double) currentHeight;
+
+        if (requestedHorizontalRatio > originalVerticalRatio) {
```

Comment:
It is not clear to me why this is the correct condition. I think this either needs a comment or the code should be simplified. For example, you could first calculate the width and if it is too large, reset it and calculate the height instead.